### ISSUE TITLE
Dataset: improve handling of multiple instances of the same dataset

### ISSUE THREAD

**Issue by jstriebel** [scalableminds/webknossos-libs#524]
In case a Dataset is opened multiple times, former writes of another instance would currently silently be overwritten. We should raise errors in those cases and completely prevent them where possible:
* We could cache the datasource-properties content in a dataset and assert that it's still unchanged before we rewrite the file (atomically if possible). If it's changed, we should raise an error.
* When a dataset is opened multiple times in the same thread, we might return the first dataset instead of creating a new one to ensure a shared state.

**Comment by jstriebel** [comment]
cc @normanrz

**Comment by normanrz** [comment]
A few thoughts:
* Maybe the data should not be cached at all, but always read from disk?
* `*.lock` files could help to avoid concurrent writes to the file.
* I would not optimize for the single-node multiprocess/thread case, but instead for the general slurm/cloud/multiprocess/etc case.

**Comment by jstriebel** [comment]
I'm not sure if we can control what is cached, since the user-code might re-use previous results to change the datasource-properties in latter calls. So we don't have any control over what is cached as soon as we expose any data from the datasource-properties. As an extreme example, a user might expand the bounding box of a layer by different degrees in parallel:
```
def expand_bounding_box_by_50():
    ds = …
    ds.get_layer("color").bounding_box += (50, 50, 50)
    # This is not atomic and can not be locked implicitly on our side, since there's a read and write from user-code.

def expand_bounding_box_by_100():
    ds = …  # same ds as above
    ds.get_layer("color").bounding_box += (100, 100, 100)

```

In this case, it might happen that the bounding box is expanded by 150 as expected, but also it might happen that it's just expanded by 50 or 100, e.g. using the following order:
```
read_bbox_before_50 = ds.get_layer("color").bounding_box
read_bbox_before_100 = ds.get_layer("color").bounding_box
ds.get_layer("color").bounding_box = read_bbox_before_100 + (100, 100, 100)
ds.get_layer("color").bounding_box = read_bbox_before_50 + (50, 50, 50)
```

We can make some operations atomic where all read and writes happen within one user-call (e.g. downsampling) by using locking or not caching data. But as soon as datasource-properties data is read and modified outside of the lib, we can only detect such cases and offer nicer solutions (such as allowing the user to add a lock).

Also, I think that the single-threaded multi-instance case is quite important to manage, too, and can be fixed, in contrast to the parallel case:
```
ds_for_50 = …
ds_for_100 = …  # same ds, opened a second time (e.g. due to complex function calls)
ds_for_50.get_layer("color").bounding_box += (50, 50, 50)
ds_for_100.get_layer("color").bounding_box += (100, 100, 100)
```

Having the same instance for `ds_for_100` and `ds_for_50` would be a simple fix here. In the distributed (multiprocessing, slurm, …) case we'd need to implement a shared object to allow similar logic (which the file already is) and would still have the same problem from above in the parallel access case.

So I think there are five different things we could do, and probably should do all of them:
1. Use a locking mechanism to make library-internal calls atomic, e.g. downsampling. When aquiring the lock, the state of the object must be synced to the datasource-properties, or nothing is cached in the dataset.
2. Expose this locking mechanism to the user, so that they can make their read/transform/write code atomic. Also simplify re-syncing to the disk state (e.g. after another spawned process changed the dataset).
3. Detect invalid writes to the datasource-properties (e.g. due to missing locking in user code), where invalid means that the underlying object expects a different datasource-properties than the one on disk. We might be generous when the diff is in an unrelated part of the json and can easily be reconciled (e.g. parallel writes to different layer bounding-boxes).
4. Dataset-instances pointing to the same files on disk should be the same instance where possible (in the single threaded case)
5. We could try to auto-sync datasets either using multiprocessing.Manager, and/or, more generally, watching the file, e.g. with [watchdog](https://pythonhosted.org/watchdog/api.html#watchdog.observers.Observer). However, this would not replace the locking mechanism (also exposed to the user), since read/transform/write calls are still not atomic.

**Comment by normanrz** [comment]
I was wondering, if this problem would be better solved, if we disallowed out-of-bounds writes? I stumbled upon this, because for the Zarr implementation, we need explicit resizes in order to support out-of-bounds writes.

**Comment by philippotto** [comment]
I think, this is orthogonal to this issue, since even explicit resizes could happen in parallel (leading to inconsistencies). If there are advantages of that approach, we should probably discuss this in a new issue.

**Comment by normanrz** [comment]
It may be orthogonal to this particular issue, but I think it would solve the issue we have in Voxelytics, no?

**Comment by philippotto** [comment]
The particular issue in voxelytics has already been solved with https://github.com/scalableminds/webknossos-libs/pull/610 and https://github.com/scalableminds/voxelytics/pull/2911. So, I don't see a need for requiring explicit resizes at the moment.

**Comment by jstriebel** [comment]
IMO we can close this issue now, since the detection with a helpful warning is implemented (PR #633), and all other steps seem too intrusive atm. Please re-open if you disagree @normanrz @philippotto
